### PR TITLE
Improvements around exceptions handlings and better logging

### PR DESCRIPTION
**Pull request by sbrossie**

### CODE REVIEW

File: src/test/java/org/killbill/billing/plugin/bridge/api/resolver/remote/TestRemoteResolverRequest.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+import org.killbill.billing.plugin.bridge.api.resolver.remote.RemoteResolver.WrappedKillBillClientException;
+import org.killbill.billing.plugin.bridge.api.resolver.remote.RemoteResolver.WrappedUnresolvedException;
+import org.killbill.billing.plugin.bridge.api.resolver.remote.RemoteResolverRequest.UnresolvedException;
+import org.mockito.Mockito;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class TestRemoteResolverRequest {
+
+    @Test
```
Comment (pierre) [inline src/test/java/org/killbill/billing/plugin/bridge/api/resolver/remote/TestRemoteResolverRequest.java]: Missing groups (won't run automatically).

File: src/test/java/org/killbill/billing/plugin/bridge/TestBridgeConfigConfigurationHandlerForTest.java [hunk 2]
```diff
@@ -1,40 +0,0 @@
...
-        final BridgeConfigConfigurationHandlerForTest bridgeConfigConfigurationHandlerForTest = new BridgeConfigConfigurationHandlerForTest("eu-central-1");
-        final BridgeConfig bridgeConfig = bridgeConfigConfigurationHandlerForTest.getConfigurable(UUID.randomUUID());
-
-        Assert.assertEquals(bridgeConfig.killbillClientConfig.serverUrl.toString().compareTo("http://kb.eu:8080"), 0);
-        Assert.assertNull(bridgeConfig.paymentConfig.getControlPlugins());
-        Assert.assertEquals(bridgeConfig.paymentConfig.getPluginProperties().size(), 2);
-        Assert.assertEquals(bridgeConfig.paymentConfig.getPluginProperties().get(0).getKey(), "skipGw");
-        Assert.assertEquals(bridgeConfig.paymentConfig.getPluginProperties().get(0).getValue(), true);
-        Assert.assertEquals(bridgeConfig.paymentConfig.getPluginProperties().get(1).getKey(), "deposit");
-        Assert.assertEquals(bridgeConfig.paymentConfig.getPluginProperties().get(1).getValue(), "{\"system\":[{\"type\":\"default\",\"corporation\":\"acme\"}]}");
```
Comment (pierre) [inline src/test/java/org/killbill/billing/plugin/bridge/TestBridgeConfigConfigurationHandlerForTest.java]: Why did you remove it? It was in place to verify the serialization of JSON properties.

File: src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolverRequest.java [hunk 3]
```diff
@@ -65,56 +66,66 @@ public RemoteResolverRequest resolveAccount(final org.killbill.billing.account.a
...
                         response.setPaymentIdMapping(payment.getPaymentId());
                         final Optional<PaymentTransaction> transaction = payment.getTransactions().stream()
-                                .filter(t -> t.getTransactionExternalKey().equals(transactionExternalKey))
-                                .findAny();
+                                                                                .filter(t -> t.getTransactionExternalKey().equals(transactionExternalKey))
+                                                                                .findAny();
                         if (transaction.isPresent()) {
                             response.setTransactionIdMapping(transaction.get().getTransactionId());
+                        } else {
+                            throw new UnresolvedException(String.format("Failed to resolve payment transaction id='%s'", transactionExternalKey));
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolverRequest.java]: `externalKey` instead of `id`.

File: src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolverRequest.java [hunk 4]
```diff
@@ -65,56 +66,66 @@ public RemoteResolverRequest resolveAccount(final org.killbill.billing.account.a
...
-    public RemoteResolverRequest resolvePayment(final String paymentExternalKey, final DefaultAction defaultAction) {
+    public RemoteResolverRequest resolvePayment(final String paymentExternalKey) throws UnresolvedException {
 
         if (paymentExternalKey != null) {
             requests.add(new Request(ResolvingType.PAYMENT, paymentExternalKey, (client, requestOptions, response) -> {
                 final Payment payment = client.getPaymentByExternalKey(paymentExternalKey, requestOptions);
                 if (payment != null) {
                     response.setPaymentIdMapping(payment.getPaymentId());
+                } else {
+                    throw new UnresolvedException(String.format("Failed to resolve payment  id='%s'", paymentExternalKey));
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolverRequest.java]: `externalKey` instead of `id`.

File: src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolverRequest.java [hunk 5]
```diff
@@ -65,56 +66,66 @@ public RemoteResolverRequest resolveAccount(final org.killbill.billing.account.a
...
+
         if (pmExternalKey != null) {
             requests.add(new Request(ResolvingType.PAYMENT_METHOD, pmExternalKey, (client, requestOptions, response) -> {
                 final PaymentMethod pm = client.getPaymentMethodByKey(pmExternalKey, requestOptions);
-                if (defaultAction == DefaultAction.CREATE_IF_MISSING && pm == null) {
-                    // TODO
+                if (pm != null) {
+                    response.setPaymentMethodIdMapping(pm.getPaymentMethodId());
+                } else {
+                    throw new UnresolvedException(String.format("Failed to resolve payment method id='%s'", pmExternalKey));
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolverRequest.java]: `externalKey` instead of `id`.

File: src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolverRequest.java [hunk 6]
```diff
@@ -65,56 +66,66 @@ public RemoteResolverRequest resolveAccount(final org.killbill.billing.account.a
                     }
                     account = client.createAccount(input, requestOptions);
                 }
-                response.setAccountIdMapping(account.getAccountId());
+                if (account.getAccountId() != null) {
+                    response.setAccountIdMapping(account.getAccountId());
+                } else {
+                    throw new UnresolvedException(String.format("Failed to resolve account id='%s'", srcAccount.getExternalKey()));
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolverRequest.java]: `externalKey` instead of `id`.

File: src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolver.java [hunk 7]
```diff
@@ -37,20 +38,44 @@ public RemoteResolver(final KillBillClient client, final RequestOptions requestO
...
-
+               .stream()
+               .forEachOrdered(r -> {
+                   boolean resolved = false;
+                   WrappedKillBillClientException clientException = null;
+                   try {
+                       r.resolve(client, requestOptions, result);
+                       resolved = true;
+                   } catch (final KillBillClientException e) {
+                       logger.warn(String.format("RemoteResolver [%d]: KillBillClientException...", request.getRequestId()), e);
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolver.java]: No need for `String.format` (`{}` notation).

File: src/test/java/org/killbill/billing/plugin/bridge/TestBridgeConfigConfigurationHandlerForTest.java (hunk 2 above)
Comment (sbrossie) [reply src/test/java/org/killbill/billing/plugin/bridge/TestBridgeConfigConfigurationHandlerForTest.java]: Mistake will fix.

File: src/test/java/org/killbill/billing/plugin/bridge/api/resolver/remote/TestRemoteResolverRequest.java (hunk 1 above)
Comment (sbrossie) [reply src/test/java/org/killbill/billing/plugin/bridge/api/resolver/remote/TestRemoteResolverRequest.java]: Ack, will add.

File: src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolver.java (hunk 7 above)
Comment (sbrossie) [reply src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolver.java]: I fixed this one, but before i got further and fix them all, a few questions:
* Is there a limit in the # of '{}' i can have ?
* This seems to work when using `logger.xx` but not when using `logService.log()` (`OSGIKillbillLogService`), which brings me to following question: what is the right way to log from a plugin?

File: src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolverRequest.java (hunk 3 above)
Comment (sbrossie) [reply src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolverRequest.java]: ack, will fix.

File: src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolverRequest.java (hunk 4 above)
Comment (sbrossie) [reply src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolverRequest.java]: ack, will fix.

File: src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolverRequest.java (hunk 6 above)
Comment (sbrossie) [reply src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolverRequest.java]: ack, will fix.

File: src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolver.java (hunk 7 above)
Comment (pierre) [reply src/main/java/org/killbill/billing/plugin/bridge/api/resolver/remote/RemoteResolver.java]: > * Is there a limit in the # of '{}' i can have ?

I don't think so, see the implementation [here](https://github.com/qos-ch/slf4j/blob/5e93462dd781677fb75aefc6cc7df65822970ca0/slf4j-api/src/main/java/org/slf4j/helpers/MessageFormatter.java#L186-L241).

> * This seems to work when using `logger.xx` but not when using `logService.log()` (`OSGIKillbillLogService`)

Indeed: `{}` is part of the SLF4J API. `OSGIKillbillLogService` implements `LogService` (OSGI spec), which only understands already-formatted messages.

> which brings me to following question: what is the right way to log from a plugin?  

Short answer: use SLF4J directly.

Longer answer: we need to use the OSGI mechanism `LogService` in plugins (e.g. to give us access to `LogEntry` events displayed in Kaui) but because the API is somewhat limited (e.g. no formatting options) and because we want logs to respect the main `logback.xml` (to support obfuscation for instance), we have a bridge in place to redirect `SLF4J` logs from plugins into the OSGI `LogService`and back into `SLF4J` in Kill Bill (https://github.com/killbill/killbill/wiki/Logback-recipes#handling-plugin-logs).
